File: mobile/lib/protocol/auth_ok_payload.dart
import 'dart:typed_data';

import 'frame.dart';
import 'payload_codec.dart';

const int authOkResultOk = 0x00;

class AuthOkPayload {
  const AuthOkPayload({
    required this.result,
    required this.sessionId,
    required this.serverCapabilities,
    required this.newToken,
  });

  final int result;
  final Uint8List sessionId;
  final int serverCapabilities;
  final String newToken;

  @override
  bool operator ==(Object other) =>
      other is AuthOkPayload &&
      result == other.result &&
      _bytesEqual(sessionId, other.sessionId) &&
      serverCapabilities == other.serverCapabilities &&
      newToken == other.newToken;

  @override
  int get hashCode => Object.hash(
        result,
        Object.hashAll(sessionId),
        serverCapabilities,
        newToken,
      );
}

class AuthOkPayloadCodec {
  static Uint8List encode(AuthOkPayload payload) {
    if (payload.result != authOkResultOk) {
      throw const ProtocolException('Unknown AUTH_OK result.');
    }
    if (payload.sessionId.length != 16) {
      throw const ProtocolException('sessionId must be exactly 16 bytes.');
    }

    final writer = PayloadWriter()
      ..writeUInt8(payload.result)
      ..writeBytes(payload.sessionId)
      ..writeUInt32(payload.serverCapabilities)
      ..writeString(payload.newToken);
    return writer.toBytes();
  }

  static AuthOkPayload decode(List<int> bytes) {
    final reader = PayloadReader(bytes);
    final result = reader.readUInt8();
    final sessionId = reader.readBytes(16);
    final serverCapabilities = reader.readUInt32();
    final newToken = reader.readString();
    reader.expectEnd();

    if (result != authOkResultOk) {
      throw const ProtocolException('Unknown AUTH_OK result.');
    }

    return AuthOkPayload(
      result: result,
      sessionId: sessionId,
      serverCapabilities: serverCapabilities,
      newToken: newToken,
    );
  }
}

bool _bytesEqual(Uint8List a, Uint8List b) {
  if (a.length != b.length) return false;
  for (var i = 0; i < a.length; i++) {
    if (a[i] != b[i]) return false;
  }
  return true;
}

File: desktop/Protocol/AuthOkPayload.cs
namespace CTRL.Desktop.Protocol;

public sealed record AuthOkPayload(
    byte Result,
    byte[] SessionId,
    uint ServerCapabilities,
    byte[] NewToken);

public static class AuthOkPayloadCodec
{
    public const byte ResultOk = 0x00;

    public static byte[] Encode(AuthOkPayload payload)
    {
        if (payload.Result != ResultOk)
            throw new ProtocolException("Unknown AUTH_OK result.");
        if (payload.SessionId is null || payload.SessionId.Length != 16)
            throw new ProtocolException("sessionId must be exactly 16 bytes.");
        if (payload.NewToken is null)
            throw new ProtocolException("newToken must not be null.");
        if (payload.NewToken.Length > byte.MaxValue)
            throw new ProtocolException("newToken must not exceed 255 bytes.");

        var writer = new PayloadWriter();
        writer.WriteUInt8(payload.Result);
        writer.WriteBytes(payload.SessionId);
        writer.WriteUInt32(payload.ServerCapabilities);
        writer.WriteUInt8((byte)payload.NewToken.Length);
        writer.WriteBytes(payload.NewToken);
        return writer.ToArray();
    }

    public static AuthOkPayload Decode(byte[] payload)
    {
        var reader = new PayloadReader(payload);
        var result = reader.ReadUInt8();
        var sessionId = reader.ReadBytes(16);
        var serverCapabilities = reader.ReadUInt32();
        var newTokenLength = reader.ReadUInt8();
        var newToken = reader.ReadBytes(newTokenLength);
        reader.ExpectEnd();

        if (result != ResultOk)
            throw new ProtocolException("Unknown AUTH_OK result.");

        return new AuthOkPayload(result, sessionId, serverCapabilities, newToken);
    }
}
